src/Stoa.ts

```diff
@@ -147,18 +168,124 @@ class Stoa {
+            {
+                if  (
+                    (block.header !== undefined) &&
+                    (block.header.height !== undefined) &&
+                    (block.header.height.value !== undefined)
+                )
+                {
+                    return Number(block.header.height.value);
+                }
+                else
+                {
+                    throw Error("Not found block height in JSON Block");
+                }
+            };
+
+            // When a new block arrives while the block is being stored,
+            // the processing of the new block is pushed back
+            // so as not to affect the previous task.
+            if (this.saving_block)
+            {
+                try
+                {
+                    let height = getBlockHeight(block);
+                    this.pushTestEvent(`put off ${height}`);
+                    console.log(`Put off a block with block height of ${height}`);
+                    res.status(200).send();
+                }
+                catch (error)
+                {
+                    console.error("Failed to save block:\n" + error + "\n" + JSON.stringify(block));
+                    res.status(500).send("An error occurred while saving");
+                }
 
-            this.ledger_storage.putBlocks(block)
-            .then(() => {
-                res.status(200).send();
                 return;
-            })
-            .catch((err) => {
-                console.error("Failed to store the payload of a push to the DB: " + err);
+            }
+
+            try
+            {
+                let height = getBlockHeight(block);
+
+                // Obtain the height of the block that should be recorded in the database.
+                let expected_height = await this.ledger_storage.getExpectedBlockHeight();
+                if (height == expected_height)
+                {
+                    // Save a block just received
+                    await this.ledger_storage.putBlocks(block);
+                    this.pushTestEvent(`save ${height}`);
+                    console.log(`Saved a block with block height of ${height}`);
+
+                    res.status(200).send();
+                }
+                else if (height > expected_height)
+                {
+                    // Recovery is required for blocks that are not received.
+                    // start is `expected_height`
+                    // count is `block_data.header.height - expected_height`
+                    // Record a current block after recovery.
+
+                    this.saving_block = true;
+
+                    let block_height = expected_height;
+                    let max_blocks = height - expected_height;
+
+                    let agora_client = new AgoraClient(this.agora_address, this.agora_port);
+                    let blocks = await agora_client.requestBlocks(block_height, max_blocks);
```

Comment:
The recovery code should not be in the handler. This could lead to infinite recursion very easily.